core/trino-main/src/main/java/io/trino/operator/scalar/JsonStringArrayExtractScalar.java

```diff
@@ -0,0 +1,159 @@
+import io.trino.spi.type.ArrayType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.TypeSignature;
+import io.trino.type.JsonPathType;
+import io.trino.util.JsonCastException;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandle;
+
+import static com.fasterxml.jackson.core.JsonToken.END_ARRAY;
+import static com.fasterxml.jackson.core.JsonToken.START_ARRAY;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.trino.spi.StandardErrorCode.INVALID_CAST_ARGUMENT;
+import static io.trino.spi.function.InvocationConvention.InvocationArgumentConvention.NEVER_NULL;
+import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.NULLABLE_RETURN;
+import static io.trino.spi.type.TypeSignature.arrayType;
+import static io.trino.spi.type.TypeSignatureParameter.typeVariable;
+import static io.trino.util.JsonUtil.createJsonFactory;
+import static io.trino.util.JsonUtil.createJsonParser;
+import static io.trino.util.JsonUtil.truncateIfNecessaryForErrorMessage;
+import static io.trino.util.Reflection.methodHandle;
+import static java.lang.String.format;
+
+public final class JsonStringArrayExtractScalar
+        extends SqlScalarFunction
+{
+    public static final JsonStringArrayExtractScalar JSON_STRING_ARRAY_EXTRACT_SCALAR = new JsonStringArrayExtractScalar();
+    public static final String JSON_STRING_ARRAY_EXTRACT_SCALAR_NAME = "$internal$json_string_array_extract_scalar";
+    private static final MethodHandle METHOD_HANDLE = methodHandle(JsonStringArrayExtractScalar.class, "extract", ArrayType.class, ArrayBlockBuilderAppender.class, ConnectorSession.class, Slice.class, JsonPath.class);
+
+    private static final JsonFactory JSON_FACTORY = createJsonFactory();
+
+    static {
+        // Changes factory. Necessary for JsonParser.readValueAsTree to work.
+        new ObjectMapper(JSON_FACTORY);
+    }
+
+    private JsonStringArrayExtractScalar()
+    {
+        super(FunctionMetadata.scalarBuilder(JSON_STRING_ARRAY_EXTRACT_SCALAR_NAME)
+                .signature(Signature.builder()
+                        .argumentType(new TypeSignature("varchar", typeVariable("N")))
+                        .longVariable("N")
+                        .argumentType(new TypeSignature(JsonPathType.NAME))
+                        .returnType(arrayType(new TypeSignature("varchar", typeVariable("N"))))
+                        .build())
+                .nullable()
+                .hidden()
+                .noDescription()
+                .build());
+    }
+
+    @Override
+    protected SpecializedSqlScalarFunction specialize(BoundSignature boundSignature)
+    {
+        checkArgument(boundSignature.getArity() == 2, "Expected arity to be 2");
+        ArrayType arrayType = (ArrayType) boundSignature.getReturnType();
+        ArrayBlockBuilderAppender arrayAppender = new ArrayBlockBuilderAppender(new VarcharBlockBuilderAppender(arrayType.getElementType()));
```

Comment:
I'm not sure why it's so complicated. You probably don't need to bind anything, as all these elements could be static:
* ArrayType is really `ArrayType(VARCHAR)`
* `VarcharBlockBuilderAppender` arg type is known (unbounded varchar)
* same for `ArrayBlockBuilderAppender`

you could essentially make static methods out of these classes

Reply:
However, if were to support `transform(cast(json_parse(varchar_col) as array(json)), json -> cast(json_extract_scalar(..) as X))` then these bounded variables would be useful.

Have you seen extra cast on top of `json_extract_scalar`?

Reply:
No such situation observed. Removed complicated binding.